src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationPair.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        private partial class CompilationTracker
+        {
+            /// <summary>
+            /// When we're working with compilations, we often have two: a compilation that does not contain generated files
+            /// (which we might need later to run generators again), and one that has the stale generated files that we might
+            /// be able to reuse as well. In those cases we have to do the same transformations to both, and this gives us
+            /// a handy way to do precisely that while not forking compilations twice if there are no generated files anywhere.
+            /// </summary>
+            internal struct CompilationPair
```

Comment:
This should also be reusable so we can clean up some of the duplication in FinalizeCompilation; I'm leaving that out of scope _for this PR_ since this is targeting 17.0, and I don't want to churn more than I have to. In 17.1, let's go for it.

Reply:
readonly struct?